packages/ngtools/webpack/src/ivy/plugin.ts

```diff
@@ -0,0 +1,489 @@
+    warnings,
+    errors,
+    compiler.context,
+    tsconfig,
+  );
+
+  return { processor, errors, warnings };
+}
+
+const PLUGIN_NAME = 'angular-compiler';
+
+export class AngularWebpackPlugin {
+  private readonly pluginOptions: AngularPluginOptions;
+  private watchMode?: boolean;
+  private ngtscNextProgram?: ts.Program;
+  private builder?: ts.EmitAndSemanticDiagnosticsBuilderProgram;
+  private sourceFileCache?: Map<string, ts.SourceFile>;
+  private buildTimestamp!: number;
+
+  constructor(options: Partial<AngularPluginOptions> = {}) {
+    this.pluginOptions = {
+      emitClassMetadata: false,
+      emitNgModuleScope: false,
+      fileReplacements: {},
+      substitutions: {},
+      directTemplateLoading: true,
+      tsconfig: 'tsconfig.json',
+      ...options,
+    };
+  }
+
+  get options() {
+    return this.pluginOptions;
+  }
+
+  apply(compiler: Compiler & { watchMode?: boolean }): void {
+    // Setup file replacements with webpack
+    for (const [key, value] of Object.entries(this.pluginOptions.fileReplacements)) {
+      new NormalModuleReplacementPlugin(
+        new RegExp('^' + key.replace(/[.*+\-?^${}()|[\]\\]/g, '\\$&') + '$'),
+        value,
+      ).apply(compiler);
+    }
+
+    // Mimic VE plugin's systemjs module loader resource location for consistency
+    new ContextReplacementPlugin(
+      /\@angular[\\\/]core[\\\/]/,
+      path.join(compiler.context, '$$_lazy_route_resource'),
+    ).apply(compiler);
+
+    // Set resolver options
+    const pathsPlugin = new TypeScriptPathsPlugin();
+    compiler.hooks.afterResolvers.tap('angular-compiler', (compiler) => {
+      // When Ivy is enabled we need to add the fields added by NGCC
+      // to take precedence over the provided mainFields.
+      // NGCC adds fields in package.json suffixed with '_ivy_ngcc'
+      // Example: module -> module__ivy_ngcc
+      // tslint:disable-next-line:no-any
+      (compiler as any).resolverFactory.hooks.resolveOptions
+        .for('normal')
+        .tap(PLUGIN_NAME, (resolveOptions: { mainFields: string[] }) => {
+          const originalMainFields = resolveOptions.mainFields;
+          const ivyMainFields = originalMainFields.map((f) => `${f}_ivy_ngcc`);
+
+          return {
+            ...resolveOptions,
+            mainFields: [...ivyMainFields, ...originalMainFields],
+          };
+        });
+
+      // tslint:disable-next-line:no-any
+      (compiler as any).resolverFactory.hooks.resolver
+        .for('normal')
+        .tap(PLUGIN_NAME, (resolver: {}) => {
+          pathsPlugin.apply(resolver);
+        });
+    });
+
+    let ngccProcessor: NgccProcessor | undefined;
+    const resourceLoader = new WebpackResourceLoader();
+    let previousUnused: Set<string> | undefined;
+    compiler.hooks.thisCompilation.tap(PLUGIN_NAME, (thisCompilation) => {
+      const compilation = thisCompilation as WebpackCompilation;
+
+      // Store watch mode; assume true if not present (old webpack versions)
```

Comment:
Should we still handle this?

Reply:
The peer dependency for Webpack is still `^4.0.0` and `watchMode` was added in `4.23.0`.  Once we eventually shift to Webpack 5.0 than we can probably adjust this.
I simplified the logic though so its really not much more code.